src/libs/highlevel/elektra_error.c

```diff
@@ -0,0 +1,322 @@
+
+// elektra_error_private.h
+
+/**
+ * Creates a new ElektraError using the provided values.
+ * The returned value will be allocated with elektraCalloc().
+ *
+ * @param code        The error code of the error.
+ * @param description The description of the error.
+ * @param severity    The severity of the error. Only use ELEKTRA_ERROR_SEVERITY_FATAL,
+ *                    if the error will be raised with elektraFatalError().
+ * @param group       The group to which this error belongs.
+ * @param module      The module from which this error originates.
+ * @return A newly allocated ElektraError (free with elektraFree()).
+ */
+ElektraError * elektraErrorCreate (ElektraErrorCode code, const char * description, ElektraErrorSeverity severity)
+{
+	ElektraError * const error = elektraCalloc (sizeof (struct _ElektraError));
+	error->code = code;
+	error->description = elektraStrDup (description);
+	error->severity = severity;
+
+	return error;
+}
+
+/**
+ * Creates a new ElektraError by using the values of the error metadata of a Key.
+ *
+ * @param Key The key from which the error data shall be taken.
+ * @return A new ElektraError created with elektraErrorCreate().
+ */
+ElektraError * elektraErrorCreateFromKey (Key * key)
+{
+	const Key * metaKey = keyGetMeta (key, "error");
+
+	if (NULL == metaKey)
+	{
+		return NULL;
+	}
+
+	kdb_long_t code;
+	elektraKeyToLong (keyGetMeta (key, "error/number"), &code);
+
+	const char * severityString = keyString (keyGetMeta (key, "error/severity"));
+	ElektraErrorSeverity severity = ELEKTRA_ERROR_SEVERITY_FATAL; // Default is FATAL.
+	if (!elektraStrCmp (severityString, "error"))
+	{
+		severity = ELEKTRA_ERROR_SEVERITY_ERROR;
+	}
+	else if (!elektraStrCmp (severityString, "warning"))
+	{
+		severity = ELEKTRA_ERROR_SEVERITY_WARNING;
+	}
+
+	ElektraKDBErrorGroup group = keyString (keyGetMeta (key, "error/ingroup"));
+	ElektraKDBErrorModule module = keyString (keyGetMeta (key, "error/module"));
+
+
+	ElektraError * error;
+
+	const char * description = keyString (keyGetMeta (key, "error/description"));
+	error = elektraErrorLowLevel (severity, code, description, group, module);
+
+	error->lowLevelError = elektraKDBErrorFromKey (key);
+	return error;
+}
+
+// elektra_error.h
+
+/**
+ * \addtogroup highlevel High-level API
+ * @{
+ */
+
+/**
+ * @return the error code of the given error
+ */
+ElektraErrorCode elektraErrorCode (const ElektraError * error)
+{
+	return error->code;
+}
+
+/**
+ * @return the description for the given error
+ */
+const char * elektraErrorDescription (const ElektraError * error)
+{
+	return error->description;
+}
+
+/**
+ * @return the severity of the given error
+ */
+ElektraErrorSeverity elektraErrorSeverity (const ElektraError * error)
+{
+	return error->severity;
+}
+
+/**
+ * @return the Key from which this error was extracted,
+ * or NULL if the error was created by the high-level API
+ */
+ElektraKDBError * elektraErrorLowLevelError (const ElektraError * error)
```

Comment:
domhof won't be able to help, because I wrote that part myself...

The alternative would be to put the Key*  into `ElektraError` directly, which would mean you would have to decipher the Key metadata to access more information about a low-level error. That would require knowledge of the low-level API and its conventions for storing errors.

If we decide to remove `ElektraKDBError`, I think we should at least append the reason string to the description string, because the description of low-level errors isn't always very useful.

Reply:
Having ElektraKDBError internally is perfectly ok (with getters of the individual parts of errors).